File: ios/Sources/ACCopilotExtensions/StreamingModels.swift
import Foundation

// MARK: - Streaming Models

/// Represents the different phases of streaming content.
///
/// Ported from production Teams-AdaptiveCards-Mobile SDK's streaming implementation.
public enum StreamingPhase: String, Codable, CaseIterable {
    case start = "start"
    case informative = "informative"
    case streaming = "streaming"
    case `final` = "final"

    public static func from(_ rawValue: String) -> StreamingPhase? {
        switch rawValue.lowercased() {
        case "start": return .start
        case "informative": return .informative
        case "streaming": return .streaming
        case "final": return .`final`
        default: return nil
        }
    }
}

/// Model for streaming content and state.
///
/// Contains all the metadata needed to render streaming text with proper
/// typing animations, progress indicators, and stop controls.
public struct StreamingContent: Codable, Equatable {
    /// Unique message identifier for tracking streaming state
    public let messageID: String
    /// Current phase of the streaming process
    public let phase: String
    /// The actual text content being streamed
    public let content: String
    /// Whether the streaming is complete
    public let isComplete: Bool
    /// Reason the stream ended (e.g., "done", "stopped", "error")
    public let streamEndReason: String?
    /// Characters per second for typing animation
    public let typingSpeed: Double?
    /// Whether to show a stop/cancel button
    public let showStopButton: Bool?
    /// Whether to show a progress indicator
    public let showProgressIndicator: Bool?

    /// Parsed streaming phase
    public var streamingPhase: StreamingPhase? {
        StreamingPhase.from(phase)
    }

    public init(
        messageID: String,
        phase: String,
        content: String,
        isComplete: Bool,
        streamEndReason: String? = nil,
        typingSpeed: Double? = nil,
        showStopButton: Bool? = nil,
        showProgressIndicator: Bool? = nil
    ) {
        self.messageID = messageID
        self.phase = phase
        self.content = content
        self.isComplete = isComplete
        self.streamEndReason = streamEndReason
        self.typingSpeed = typingSpeed
        self.showStopButton = showStopButton
        self.showProgressIndicator = showProgressIndicator
    }
}

// MARK: - Streaming Data Parser

/// Helper for parsing streaming data from text content.
///
/// Detects and parses JSON-encoded streaming data embedded in text elements.
/// This allows regular TextBlock elements to carry streaming metadata.
public struct StreamingDataParser {

    /// Attempts to parse streaming data from a text string
    public static func parseStreamingData(from text: String) -> StreamingContent? {
        guard text.hasPrefix("{") && text.hasSuffix("}"),
              let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              json["streamingEnabled"] as? Bool == true,
              let messageID = json["messageID"] as? String,
              let phase = json["phase"] as? String,
              let content = json["content"] as? String else {
            return nil
        }

        return StreamingContent(
            messageID: messageID,
            phase: phase,
            content: content,
            isComplete: json["isComplete"] as? Bool ?? false,
            streamEndReason: json["streamEndReason"] as? String,
            typingSpeed: json["typingSpeed"] as? Double,
            showStopButton: json["showStopButton"] as? Bool,
            showProgressIndicator: json["showProgressIndicator"] as? Bool
        )
    }

    /// Checks if text content contains valid streaming data
    public static func isStreamingContent(_ text: String) -> Bool {
        parseStreamingData(from: text) != nil
    }
}

extension StreamingContent {
    /// Create from text content (JSON string)
    public static func from(textContent: String) -> StreamingContent? {
        StreamingDataParser.parseStreamingData(from: textContent)
    }
}

File: android/ac-copilot-extensions/src/main/kotlin/com/microsoft/adaptivecards/copilot/StreamingModels.kt
package com.microsoft.adaptivecards.copilot

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Streaming content phases and models.
 *
 * Ported from production Teams-AdaptiveCards-Mobile SDK.
 */
enum class StreamingPhase(val value: String) {
    START("start"),
    INFORMATIVE("informative"),
    STREAMING("streaming"),
    FINAL("final");

    companion object {
        fun from(rawValue: String): StreamingPhase? =
            entries.find { it.value.equals(rawValue, ignoreCase = true) }
    }
}

/**
 * Model for streaming content and state.
 */
@Serializable
data class StreamingContent(
    /** Unique message identifier */
    val messageID: String,
    /** Current phase */
    val phase: String,
    /** The text content being streamed */
    val content: String,
    /** Whether streaming is complete */
    val isComplete: Boolean,
    /** Reason the stream ended */
    val streamEndReason: String? = null,
    /** Characters per second for typing animation */
    val typingSpeed: Double? = null,
    /** Whether to show a stop button */
    val showStopButton: Boolean? = null,
    /** Whether to show a progress indicator */
    val showProgressIndicator: Boolean? = null
) {
    val streamingPhase: StreamingPhase?
        get() = StreamingPhase.from(phase)

    companion object {
        fun from(textContent: String): StreamingContent? =
            StreamingDataParser.parseStreamingData(textContent)
    }
}

/**
 * Helper for parsing streaming data from text content.
 */
object StreamingDataParser {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    fun parseStreamingData(text: String): StreamingContent? {
        if (!text.startsWith("{") || !text.endsWith("}")) return null

        return try {
            val element = json.parseToJsonElement(text)
            val obj = element.jsonObject

            val streamingEnabled = obj["streamingEnabled"]?.jsonPrimitive?.boolean ?: false
            if (!streamingEnabled) return null

            val messageID = obj["messageID"]?.jsonPrimitive?.contentOrNull ?: return null
            val phase = obj["phase"]?.jsonPrimitive?.contentOrNull ?: return null
            val content = obj["content"]?.jsonPrimitive?.contentOrNull ?: return null

            StreamingContent(
                messageID = messageID,
                phase = phase,
                content = content,
                isComplete = obj["isComplete"]?.jsonPrimitive?.boolean ?: false,
                streamEndReason = obj["streamEndReason"]?.jsonPrimitive?.contentOrNull,
                typingSpeed = try { obj["typingSpeed"]?.jsonPrimitive?.double } catch (_: Exception) { null },
                showStopButton = try { obj["showStopButton"]?.jsonPrimitive?.boolean } catch (_: Exception) { null },
                showProgressIndicator = try { obj["showProgressIndicator"]?.jsonPrimitive?.boolean } catch (_: Exception) { null }
            )
        } catch (_: Exception) {
            null
        }
    }

    fun isStreamingContent(text: String): Boolean = parseStreamingData(text) != null
}
